lib/Optimizer/Transforms/MemToReg.cpp

```diff
@@ -820,325 +678,344 @@ class MemToRegPass : public cudaq::opt::impl::MemToRegBase<MemToRegPass> {
-              auto pr = dataFlow.addEscapingBinding(block, memdef);
-              if (pr.second)
-                appendPredecessorsToWorklist(worklist, block);
             }
+            cleanUps.insert(defop);
+          };
+          if (auto wrap = dyn_cast<quake::WrapOp>(op)) {
+            if (quantumValues)
+              handleDefinition(wrap, wrap.getWireValue(), wrap.getRefValue());
+            continue;
           }
-          dataFlow.addBinding(block, memdef, val);
-        };
-        if (auto wrap = dyn_cast<quake::WrapOp>(op)) {
-          if (quantumValues)
-            handleDefinition(wrap, wrap.getWireValue(), wrap.getRefValue());
-          continue;
-        }
-        if (auto store = dyn_cast<cudaq::cc::StoreOp>(op)) {
-          if (classicalValues) {
-            auto memdef = store.getPtrvalue();
-            // Process only singleton classical scalars, no aggregates.
-            if (auto *defOp = memdef.getDefiningOp())
-              if (memAnalysis.isMember(defOp))
-                handleDefinition(store, store.getValue(), store.getPtrvalue());
+          if (auto store = dyn_cast<cudaq::cc::StoreOp>(op)) {
+            if (classicalValues) {
+              auto memdef = store.getPtrvalue();
+              // Process only singleton classical scalars, no aggregates.
+              if (auto *defOp = memdef.getDefiningOp())
+                if (memAnalysis.isMember(defOp))
+                  handleDefinition(store, store.getValue(),
+                                   store.getPtrvalue());
+            }
+            continue;
           }
-          continue;
+
+          // If op uses a quantum reference, then halt forwarding the unwrap
+          // use chain and leave a wrap dominating op.
+          for (auto v : op->getOperands())
+            if ((v.getType() == qrefTy) && dataFlow.hasBinding(block, v))
+              if (auto vBinding = dataFlow.getBinding(block, v)) {
+                OpBuilder builder(op);
+                builder.create<quake::WrapOp>(op->getLoc(), vBinding, v);
+                dataFlow.cancelBinding(block, v);
+              }
+
+        } // end loop over ops
+      }   // end loop over blocks
+    }     // end loop over regions
+
+    LLVM_DEBUG(llvm::dbgs() << "After threading intra-block:\n"
+                            << *parent << "\n\n");
+
+    std::deque<Block *> worklist;
+    appendToWorklist(worklist, dataFlow.getExitBlocks());
+
+    // 3. If there are defs that are live-out for parent and parent takes region
+    // arguments, construct a list of live-in region arguments to add to the new
+    // parent and replace uses of promoted defs with block arguments.
+    dataFlow.updatePromotedDefs(parent, worklist);
+
+    LLVM_DEBUG(llvm::dbgs() << "After fixing up promoted loads:\n"
+                            << *parent << "\nPromotions:\n");
+    for (auto v : dataFlow.getPromotedDefValues()) {
+      LLVM_DEBUG(v.dump());
+    }
+    LLVM_DEBUG(llvm::dbgs() << '\n');
```

Comment:
If the following is equivalent (double check me), it would avoid iterating over the loop if debug was not specified.
```suggestion
    LLVM_DEBUG({llvm::dbgs() << "After fixing up promoted loads:\n"
                             << *parent << "\nPromotions:\n";
    for (auto v : dataFlow.getPromotedDefValues()) {
      v.dump();
    }
    llvm::dbgs() << '\n'});
```